# 项目：评估和清理英国电商公司销售数据

## 分析目标

此数据分析的目的是，根据市场销售数据，挖掘畅销产品，以便制定更有效的市场策略来提升营收。

本实战项目的目的在于练习评估数据干净和整洁度，并且基于评估结果，对数据进行清洗，从而得到可供下一步分析的数据。

## 简介

原始数据集记录了一家英国在线零售公司在2010年12月1日至2011年12月9日期间的所有交易情况，涵盖了该公司在全球不同国家和地区的业务数据。该公司主要销售覆盖各个场景的礼品，包括但不限于生日礼品、结婚纪念品、圣诞礼品等等。该公司的客户群体主要包括批发商和个人消费者，其中批发商占据了相当大的比例。

数据每列的含义如下：
- `InvoiceNo`: 发票号码。6位数，作为交易的唯一标识符。如果这个代码以字母“c”开头，表示这笔交易被取消。
- `StockCode`: 产品代码。5位数，作为产品的唯一标识符。
- `Description`: 产品名称。
- `Quantity`: 产品在交易中的数量。
- `InvoiceDate`: 发票日期和时间。交易发生的日期和时间。
- `UnitPrice`: 单价。价格单位为英镑（£）。
- `CustomerID`: 客户编号。5位数，作为客户的唯一标识符。
- `Country`: 国家名称。客户所居住的国家的名称。

## 读取数据

导入数据分析所需要的库，并通过Pandas的`read_csv`函数，将原始数据文件"e_commerce.csv"里的数据内容，解析为DataFrame，并赋值给变量`original_data`。

In [1]:
import pandas as pd

In [2]:
original_data=pd.read_csv(r"C:\Users\Lenovo\Desktop\评估和清理英国电商公司销售数据项目\e_commerce.csv",encoding="gbk",low_memory=False)

In [3]:
original_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

## 评估数据

在这一部分，我将对在上一部分建立的original_data这个DataFrame所包含的数据进行评估。

评估主要从两个方面进行：结构和内容，即整齐度和干净度。数据的结构性问题指不符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”这三个标准，数据的内容性问题包括存在丢失数据、重复数据、无效数据等。

### 评估数据整齐度

In [4]:
original_data.sample(10)

InvoiceNo StockCode                      Description  Quantity  \
28683     538652     21145      ANTIQUE GLASS PLACE SETTING        24   
136195    547939     22423         REGENCY CAKESTAND 3 TIER        10   
114266    546008     22260       FELT EGG COSY BLUE RABBIT          1   
194262    564962     22748        POPPY'S PLAYHOUSE KITCHEN         1   
16354     537650     21499               BLUE POLKADOT WRAP        25   
181411    563909     22951  60 CAKE CASES DOLLY GIRL DESIGN         2   
103775    545078    85186C                BUNNY EGG GARLAND         1   
37212     539453     22109     FULL ENGLISH BREAKFAST PLATE         2   
186707    564356     20725          LUNCH BAG RED RETROSPOT         7   
137884    548171     22262            FELT EGG COSY CHICKEN        36   

             InvoiceDate  UnitPrice  CustomerID         Country  
28683   12/13/2010 15:12       1.25     17890.0  United Kingdom  
136195   3/28/2011 14:42      12.75     15611.0  United Kingdom  
114266    3/8/2011 15:43       1.63         NaN  United Kingdom  
194262   8/31/2011 14:35       2.10     17777.0  United Kingdom  
16354    12/7/2010 15:48       0.42     16125.0  United Kingdom  
181411   8/21/2011 12:36       0.55     15701.0  United Kingdom  
103775   2/28/2011 10:15       0.83         NaN  United Kingdom  
37212   12/17/2010 17:08       7.62         NaN  United Kingdom  
186707   8/24/2011 15:47       1.65     13969.0  United Kingdom  
137884   3/29/2011 13:58       0.85     16422.0  United Kingdom

从抽样的10行数据数据来看，数据符合“每列是一个变量，每行是一个观察值，每个单元格是一个值”，具体来看每行是关于某商品的一次交易，每列是交易相关的各个变量，因此不存在结构性问题。

### 评估数据干净度

In [5]:
original_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200584 entries, 0 to 200583
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   InvoiceNo    200584 non-null  object 
 1   StockCode    200584 non-null  object 
 2   Description  199786 non-null  object 
 3   Quantity     200584 non-null  int64  
 4   InvoiceDate  200584 non-null  object 
 5   UnitPrice    200584 non-null  float64
 6   CustomerID   142885 non-null  float64
 7   Country      200584 non-null  object 
dtypes: float64(2), int64(1), object(5)
memory usage: 12.2+ MB


从输出结果来看，数据共有200584条观察值，而Description、CustomerID变量存在缺失值。

此外，InvoiceDate的数据类型应为日期，应当进行数据格式转换。

### 评估缺失数据

在了解Description存在缺失值后，根据条件提取出缺失观察值。

In [6]:
original_data[original_data["Description"].isnull()]

InvoiceNo StockCode Description  Quantity      InvoiceDate  UnitPrice  \
622       536414     22139         NaN        56  12/1/2010 11:52        0.0   
1970      536545     21134         NaN         1  12/1/2010 14:32        0.0   
1971      536546     22145         NaN         1  12/1/2010 14:33        0.0   
1972      536547     37509         NaN         1  12/1/2010 14:33        0.0   
1987      536549    85226A         NaN         1  12/1/2010 14:34        0.0   
...          ...       ...         ...       ...              ...        ...   
197708    565319     23348         NaN       225   9/2/2011 13:07        0.0   
198014    565368     21784         NaN         1   9/2/2011 14:59        0.0   
198015    565369     37342         NaN         1   9/2/2011 15:02        0.0   
198019    565376     21682         NaN       144   9/2/2011 15:10        0.0   
198275    565389    47343A         NaN       -16   9/2/2011 16:19        0.0   

        CustomerID         Country  
622            NaN  United Kingdom  
1970           NaN  United Kingdom  
1971           NaN  United Kingdom  
1972           NaN  United Kingdom  
1987           NaN  United Kingdom  
...            ...             ...  
197708         NaN  United Kingdom  
198014         NaN  United Kingdom  
198015         NaN  United Kingdom  
198019         NaN  United Kingdom  
198275         NaN  United Kingdom  

[798 rows x 8 columns]

有798条交易数据缺失Description变量值。

从输出结果来看，这些缺失Description的交易数据，UnitPrice都为0。为了验证猜想，我们增加筛选条件，看是否存在Description变量缺失且UnitePrice不为0的数据

In [7]:
original_data[(original_data["Description"].isnull())&(original_data["UnitPrice"]!=0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

筛选出来结果数量为0条，说明缺失Description值的数据，同时也不具备有效的UnitPrice值。

Description表示产品名称，UnitPrice表示产品单价，都是进行后续商品交易分析的重要变量。如果它们同时缺失/无效，我们认为数据无法提供有效含义，因此这些后续可以被删除。

CustomerID变量同样存在缺失值，因此也根据条件提取出缺失观察值。

In [8]:
original_data[original_data["CustomerID"].isnull()]

InvoiceNo StockCode                      Description  Quantity  \
622       536414     22139                              NaN        56   
1443      536544     21773  DECORATIVE ROSE BATHROOM BOTTLE         1   
1444      536544     21774  DECORATIVE CATS BATHROOM BOTTLE         2   
1445      536544     21786               POLKADOT RAIN HAT          4   
1446      536544     21787            RAIN PONCHO RETROSPOT         2   
...          ...       ...                              ...       ...   
199918    565455     22203           MILK PAN RED RETROSPOT         1   
199919    565455     21246           RED RETROSPOT BIG BOWL         1   
199920    565455     23427           STOOL HOME SWEET HOME          1   
199921    565455     23424          GINGHAM RECIPE BOOK BOX         1   
199922    565455     22427          ENAMEL FLOWER JUG CREAM         1   

            InvoiceDate  UnitPrice  CustomerID         Country  
622     12/1/2010 11:52       0.00         NaN  United Kingdom  
1443    12/1/2010 14:32       2.51         NaN  United Kingdom  
1444    12/1/2010 14:32       2.51         NaN  United Kingdom  
1445    12/1/2010 14:32       0.85         NaN  United Kingdom  
1446    12/1/2010 14:32       1.66         NaN  United Kingdom  
...                 ...        ...         ...             ...  
199918   9/4/2011 16:09       3.75         NaN  United Kingdom  
199919   9/4/2011 16:09       4.95         NaN  United Kingdom  
199920   9/4/2011 16:09      12.50         NaN  United Kingdom  
199921   9/4/2011 16:09       4.95         NaN  United Kingdom  
199922   9/4/2011 16:09       5.95         NaN  United Kingdom  

[57699 rows x 8 columns]

CustomerID表示客户编号，不是分析畅销商品的必要变量。并且从输出结果来看，有些CustomerID缺失的交易数据仍然有效，因此保留此变量为空的观察值。

### 评估重复数据

根据数据变量的含义来看，虽然InvoiceNo、StockCode和CustomerID都是唯一标识符，但一次交易可能包含多件商品，因此InvoiceNo可以存在重复，不同交易可以包含同件商品，因此StockCode可以存在重复，顾客可以进行多次交易或下单多个商品，因此CustomerID也可以存在重复。

那么针对此数据集，我们无需评估重复数据。

### 评估不一致数据

不一致数据可能存在于Country变量中，我们要查看是否存在多个不同值指代同一国家的情况。

In [9]:
original_data["Country"].value_counts()

Country
United Kingdom          184065
Germany                   3580
France                    2919
EIRE                      2411
Spain                     1158
Netherlands                757
Belgium                    677
Portugal                   510
Australia                  497
Switzerland                493
Channel Islands            378
Finland                    331
Cyprus                     306
Italy                      294
Japan                      241
Norway                     232
Israel                     209
UK                         209
Sweden                     192
China                      140
Singapore                  118
Poland                     116
Unspecified                105
Austria                    104
Greece                      85
Iceland                     84
United Arab Emirates        67
Denmark                     59
Malta                       57
Lebanon                     45
Lithuania                   35
Brazil                      32


从Country变量来看，USA、United States均在表示美国，UK、United Kingdom均在表示英国，因此应该在清洗步骤对这些值进行统一，只保留一个指代值。

### 评估无效和错误数据

可以通过DataFrame的describe方法，对数值统计信息进行快速了解。

In [10]:
original_data.describe()

Quantity      UnitPrice     CustomerID
count  200584.000000  200584.000000  142885.000000
mean        9.372632       4.980377   15283.338167
std       240.348679      95.855777    1722.400894
min    -74215.000000       0.000000   12346.000000
25%         1.000000       1.250000   13869.000000
50%         3.000000       2.100000   15194.000000
75%        10.000000       4.210000   16791.000000
max     74215.000000   16888.020000   18283.000000

从输出结果来看，Quantity存在负数，会对后续数据分析产生影响。

因此，我们先筛选出Quantity数值为负数的观察值，进一步评估其意义。

In [11]:
original_data[original_data["Quantity"]<0]

InvoiceNo StockCode                          Description  Quantity  \
141      C536379         D                             Discount        -1   
154      C536383    35004C      SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556       PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984     PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983     PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                                  ...       ...   
199156   C565428     22230        JIGSAW TREE WITH WATERING CAN        -1   
199157   C565428     84755  COLOUR GLASS T-LIGHT HOLDER HANGING        -2   
199158   C565428    79190B             RETRO PLASTIC POLKA TRAY        -1   
199159   C565428         M                               Manual        -1   
200331   C565477     21012        ANTIQUE ALL GLASS CANDLESTICK        -3   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
199156   9/4/2011 12:53       0.78     12867.0  United Kingdom  
199157   9/4/2011 12:53       0.65     12867.0  United Kingdom  
199158   9/4/2011 12:53       0.42     12867.0  United Kingdom  
199159   9/4/2011 12:53       4.25     12867.0  United Kingdom  
200331   9/5/2011 10:58       2.10     13630.0  United Kingdom  

[4250 rows x 8 columns]

从筛选结果看，Quantity变量为负数的观察值，InvoiceNo似乎均以C开头，表示订单被取消。

但此猜测不一定准确，为了验证猜想，我们增加筛选条件，看是否存在Quantity变量为负数且InvoiceNo不以C开头的观察值。

In [12]:
original_data[(original_data["Quantity"]<0)&(original_data["InvoiceNo"].str[0]!="C")]

InvoiceNo StockCode  Description  Quantity      InvoiceDate  UnitPrice  \
2406      536589     21777          NaN       -10  12/1/2010 16:50        0.0   
4347      536764    84952C          NaN       -38  12/2/2010 14:42        0.0   
7188      536996     22712          NaN       -20  12/3/2010 15:30        0.0   
7189      536997     22028          NaN       -20  12/3/2010 15:30        0.0   
7190      536998     85067          NaN        -6  12/3/2010 15:30        0.0   
...          ...       ...          ...       ...              ...        ...   
197637    565314    84660C          NaN       -21   9/2/2011 12:43        0.0   
198016    565370    82494L  crushed ctn       -33   9/2/2011 15:04        0.0   
198017    565371     82483      crushed       -13   9/2/2011 15:04        0.0   
198168    565385     22435     ?missing      -434   9/2/2011 15:46        0.0   
198275    565389    47343A          NaN       -16   9/2/2011 16:19        0.0   

        CustomerID         Country  
2406           NaN  United Kingdom  
4347           NaN  United Kingdom  
7188           NaN  United Kingdom  
7189           NaN  United Kingdom  
7190           NaN  United Kingdom  
...            ...             ...  
197637         NaN  United Kingdom  
198016         NaN  United Kingdom  
198017         NaN  United Kingdom  
198168         NaN  United Kingdom  
198275         NaN  United Kingdom  

[607 rows x 8 columns]

以上猜想被证实错误，因为还存在Quantity变量为负数且InvoiceNo不以C开头的观察值。

但根据以上输出结果，这些筛选出的观察值的UnitPrice的观察值均为0，因此增加UnitPrice的条件进行筛选。

In [13]:
original_data[(original_data["Quantity"]<0) & (original_data["InvoiceNo"].str[0]!="C") & (original_data["UnitPrice"]!=0)]

Empty DataFrame
Columns: [InvoiceNo, StockCode, Description, Quantity, InvoiceDate, UnitPrice, CustomerID, Country]
Index: []

根据输出结果，猜想得到验证，当Quantity变量为负数时，观察值满足以下条件之一：

- InvoiceNo以C开头，表示订单被取消
- UnitPrice为0，说明单价是0英镑

这些交易数据均不是有效成交数据，不贡献销售，不在后续分析范围内，因此我们将在数据清洗步骤，将Quantity为负数的观察值删除

## 清理数据

根据前面评估部分得到的结论，我们需要进行的数据清理包括：

- 把InvoiceDate变量的数据类型转换为日期时间
- 把Description变量缺失的观察值删除
- 把Country变量值USA替换为United States
- 把Country变量值UK替换为United Kingdom
- 把Quantity变量值为负数的观察值删除

为了区分开经过清理的数据和原始的数据，我们创建新的变量cleaned_data，让它为original_data复制出的副本，我们之后的清理步骤都将被运用在cleaned_data上。

In [14]:
cleaned_data=original_data.copy()
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

把InvoiceDate变量的数据类型转换为日期时间:

In [16]:
cleaned_data["InvoiceDate"] = pd.to_datetime(cleaned_data["InvoiceDate"])
cleaned_data["InvoiceDate"]

0        2010-12-01 08:26:00
1        2010-12-01 08:26:00
2        2010-12-01 08:26:00
3        2010-12-01 08:26:00
4        2010-12-01 08:26:00
                 ...        
200579   2011-09-05 12:00:00
200580   2011-09-05 12:00:00
200581   2011-09-05 12:00:00
200582   2011-09-05 12:00:00
200583   2011-09-05 12:00:00
Name: InvoiceDate, Length: 200584, dtype: datetime64[ns]

把Description变量缺失的观察值删除，并查看删除后该列空缺值的个数和:

In [20]:
cleaned_data = cleaned_data.dropna(subset=["Description"],axis=0)
cleaned_data["Description"].isnull().sum()

np.int64(0)

把Country变量值USA替换为United States，并检查替换后USA变量个数:

In [38]:
cleaned_data["Country"] = cleaned_data["Country"].replace("USA","United States")    
len(cleaned_data[cleaned_data["Country"] == "USA"])

0

把Country变量值UK替换为United Kingdom，并检查替换后UK变量个数:

In [39]:
cleaned_data["Country"] = cleaned_data["Country"].replace("UK","United Kingdom")
len(cleaned_data[cleaned_data["Country"] == "UK"])

0

把Quantity变量值为负数的观察值删除，并检查替换后Quantity变量值为负数的个数:

In [40]:
cleaned_data = cleaned_data[cleaned_data["Quantity"]>=0]

In [41]:
len(cleaned_data[cleaned_data["Quantity"]<0])

0

## 保存清理后的数据

完成数据清理后，把干净整齐的数据保存到新的文件里，文件名为e_commerce_cleaned.csv

In [34]:
cleaned_data.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

          InvoiceDate  UnitPrice  CustomerID         Country  
0 2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2 2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3 2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4 2010-12-01 08:26:00       3.39     17850.0  United Kingdom

In [42]:
cleaned_data.to_csv("e_commerce_cleaned.csv",index=False)   

In [43]:
pd.read_csv("e_commerce_cleaned.csv").head()

InvoiceNo StockCode                          Description  Quantity  \
0     536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1     536365     71053                  WHITE METAL LANTERN         6   
2     536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3     536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4     536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

           InvoiceDate  UnitPrice  CustomerID         Country  
0  2010-12-01 08:26:00       2.55     17850.0  United Kingdom  
1  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
2  2010-12-01 08:26:00       2.75     17850.0  United Kingdom  
3  2010-12-01 08:26:00       3.39     17850.0  United Kingdom  
4  2010-12-01 08:26:00       3.39     17850.0  United Kingdom